python-lib/api_calling_utils.py

```diff
@@ -0,0 +1,203 @@
+    - handles errors from the function with two methods:
+        * (fail - default) fail if there is an error
+        * (warn) do not fail on a list of API-related exceptions, just log it
+        and return the row with a new 'error' key
+     """
+    assert error_handling in [i.value for i in ErrorHandlingEnum]
+
+    def inner_decorator(func):
+        if "row" not in inspect.getfullargspec(func).args:
+            raise ValueError("Function must have 'row' as first parameter.")
+
+        @wraps(func)
+        def wrapped(row, *args, **kwargs):
+            if not isinstance(row, dict):
+                raise ValueError("The 'row' parameter must be a dict.")
+            response_key = generate_unique("raw_response", row.keys())
+            error_message_key = generate_unique("error_message", row.keys())
+            error_type_key = generate_unique("error_type", row.keys())
+            error_raw_key = generate_unique("error_raw", row.keys())
+            new_keys = [
+                response_key, error_message_key,
+                error_type_key, error_raw_key
+            ]
+            if error_handling == ErrorHandlingEnum.FAIL.value:
+                row[response_key] = func(row=row, *args, **kwargs)
+                return row
+            elif error_handling == ErrorHandlingEnum.WARN.value:
+                for k in new_keys:
+                    row[k] = ''
+                try:
+                    row[response_key] = func(row=row, *args, **kwargs)
+                    return row
+                except api_exceptions as e:
+                    error_str = str(e)
+                    logging.warning(error_str)
+                    module = str(inspect.getmodule(e).__name__)
+                    class_name = str(type(e).__qualname__)
+                    error_type = module + "." + class_name
+                    error_raw = str(e.args)
+                    row[error_message_key] = error_str
+                    row[error_type_key] = error_type
+                    if verbose:
+                        row[error_raw_key] = error_raw
+                    else:
+                        del row[error_raw_key]
+                    return row
+
+        return wrapped
+
+    return inner_decorator
+
+
+def fail_or_warn_on_batch():
+    # TODO write it ;)
+    # May need to write cloud-specific decorator
+    return None
+
+
+def api_parallelizer(
+    input_df: pd.DataFrame,
+    api_call_function: Callable,
+    parallel_workers: int = 5,
+    api_support_batch: bool = False,
+    batch_size: int = 10,
+    verbose: bool = False,
+    **api_call_function_kwargs
+) -> pd.DataFrame:
+    """
+    Apply an API call function in parallel to a pandas.DataFrame.
+    The DataFrame is passed to the function as row dictionaries.
+    Parallelism works by:
+    - (default) sending multiple concurrent threads
+    - if the API supports it, sending batches of row
+    """
+    df_iterator = (i[1].to_dict() for i in input_df.iterrows())
```

Comment:
Small optimization hint: we don't need to create df_iterator as we can iterate directly over input_df in the ThreadPoolExecutor block for the non batch case.

Reply:
for this one, I am not sure... iterrows returns a tuple of index, pd.Series. So ThreadPoolExecutor needs some extra transformation to take iterrows output. I also need to have an iterator object declared to chunk it in the batch case.

What is the impact on performance to declare this iterator? It is only iterated when called, right? That's why I created an iterator comprehension instead of a list comprehension.

Reply:
Sorry I missed the iterator comprehesion, so that's good for me